metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractClass.scala

```diff
@@ -0,0 +1,250 @@
+  }
+
+  private def getNewFileTreeStructure(
+      tree: Tree,
+      range: l.Range,
+      currFileName: String
+  ): Option[(Tree, Defn.Class)] = {
+
+    val sealedDefn = tree.collect {
+      case node: Defn.Trait if node.mods.exists(_.isInstanceOf[Mod.Sealed]) =>
+        node.name.value
+      case node: Defn.Class if node.mods.exists(_.isInstanceOf[Mod.Sealed]) =>
+        node.name.value
+    }
+
+    def classExtendsSealed(classDefn: Defn.Class): Boolean =
+      getExtendsNames(classDefn).exists(sealedDefn.contains(_))
+
+    // List of sequential packages or imports before the class definition
+    val structure: ListBuffer[Stat] = ListBuffer()
+    var classFound: Option[Defn.Class] = None
+
+    // Using a custom traverser to avoid hitting inner classes by stopping the recursion on Defn.Class
+    object traverser extends SimpleTraverser {
+      override def apply(tree: Tree): Unit = tree match {
+        case p: Pkg if p.pos.toLSP.overlapsWith(range) =>
+          structure += Pkg(ref = p.ref, stats = Nil)
+          super.apply(p)
+        case i: Import =>
+          structure += i
+        case classNode: Defn.Class
+            if classNode.name.value != currFileName && classNode.name.pos.toLSP
+              .overlapsWith(range) && !classExtendsSealed(classNode)
+              && !classNode.mods.exists(_.isInstanceOf[Mod.Sealed]) =>
+          classFound = Some(classNode)
+          structure += classNode
+        case s: Source =>
+          super.apply(s)
+        case _ =>
+      }
+    }
+
+    traverser(tree)
+
+    // Folding back the list into a single Tree
+    val sourceAST: List[Stat] = structure.toList.foldRight[List[Stat]](Nil) {
+      case (p: Pkg, prev) => List(Pkg(ref = p.ref, stats = prev))
+      case (a, b) => a :: b
+    }
+
+    classFound match {
+      case Some(classDefn) => Some((Source(sourceAST), classDefn))
+      case _ => None
+    }
+  }
+
+  private def getExtendsNames(classDefn: Defn.Class): List[String] =
+    classDefn.templ.inits.flatMap {
+      _.tpe match {
+        case Type.Name(value) => Some(value)
+        case _ => None
+      }
+    }
+
+  private def renameFileAsClassAction(
+      uri: String,
+      classDefn: Defn.Class
+  ): l.CodeAction = {
+    val className = classDefn.name.value
+    val newUri = newURIFromClass(uri, classDefn)
+    val fileName = uri.toAbsolutePath.filename
+
+    val edits: List[Either[l.TextDocumentEdit, l.ResourceOperation]] = List(
+      Right(new l.RenameFile(uri, newUri))
+    )
+
+    val codeAction = new l.CodeAction()
+    codeAction.setTitle(
+      ExtractClass.renameFileAsClassTitle(fileName, className)
+    )
+    codeAction.setKind(l.CodeActionKind.Refactor)
+    codeAction.setEdit(
+      new l.WorkspaceEdit(edits.map(_.asJava).asJava)
+    )
+
+    codeAction
+  }
+
+  private def extractClassAction(
+      uri: String,
+      tree: Tree,
+      classDefn: Defn.Class
+  ): l.CodeAction = {
+
+    def versionedTextDocumentIdentifier(
```

Comment:
This should not be needed currently we don't use those anywhere

Reply:
The problem is that `TextDocumentEdit` requires a `VersionedTextDocumentIdentifier`, and the `WorkspaceEdit` requires a list of `List[Either[l.TextDocumentEdit, l.ResourceOperation]]`, so I was obliged to use `VersionedTextDocumentIdentifier`. 
But if there is another way, I'm very happy to change it! :)